src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -158,13 +159,27 @@ export class TerminalService implements ITerminalService {
 		@IEditorResolverService editorResolverService: IEditorResolverService,
 		@IExtensionService private readonly _extensionService: IExtensionService,
 		@INotificationService private readonly _notificationService: INotificationService,
+		@IWorkspaceContextService private readonly _workspaceContextService: IWorkspaceContextService,
 		@optional(ILocalTerminalService) localTerminalService: ILocalTerminalService
 	) {
 		this._localTerminalService = localTerminalService;
 		this._isShuttingDown = false;
 		this._findState = new FindReplaceState();
 		this._configHelper = _instantiationService.createInstance(TerminalConfigHelper);
-
+		this._primaryOffProcessTerminalService?.onDidRequestDetach(async (e) => {
+			if (this._workspaceContextService.getWorkspace().id === e.workspaceId) {
+				this.instances.find(i => i.instanceId === e.instanceId)?.detachFromProcess();
+			} else {
+				const processes = await this._primaryOffProcessTerminalService?.listProcesses();
+				if (processes && processes?.length > 0) {
+					const instance = this.createTerminal({
+						config: { attachPersistentProcess: processes[0] },
+						target: TerminalLocation.TerminalView
+					});
+					this.setActiveInstance(instance);
+				}
```

Comment:
We shouldn't be doing anything here, instead an event with a reply like how we resolve variables would be better such that only the workspace requesting the terminal will react to this onDidRequestDetach event:

https://github.com/microsoft/vscode/blob/b0cde4e27114c229b385c5fa3a2b9d6ed09d8d84/src/vs/platform/terminal/node/ptyHostService.ts#L312-L328

I also think there is a race condition above.

So on the pty service we would have:

```
onDidRequestDetach(workspaceId, instanceId)
acceptRequestDetachResponse(success, persistentProcessId)
```